File: heavily_right/calibration.py
"""Finite-sample covariance calibration for sample means."""

import numpy as np


def standard_error_of_mean(sample: np.ndarray, axis: int = 0) -> np.ndarray:
  """Estimate componentwise standard errors of a sample mean.

  The sample variance uses Bessel's correction, so the returned squared
  standard errors equal the diagonal of ``W / [n(n - 1)]``.

  Parameters
  ----------
  sample : array_like
      Observations; ``axis`` indexes independent sampling units.
  axis : int, default 0
      Axis removed by the mean calculation. At least two observations are
      required. Supply finite real data; this helper does not impute missingness.

  Returns
  -------
  ndarray or numpy scalar
      Standard errors in the same units as the sample values.

  Examples
  --------
  >>> import numpy as np
  >>> from heavily_right import standard_error_of_mean
  >>> sample = np.array([[1., 2.], [2., 4.], [3., 6.]])
  >>> se = standard_error_of_mean(sample)
  >>> assert np.allclose(se, [1 / np.sqrt(3), 2 / np.sqrt(3)])
  """
  sample = np.asarray(sample)
  num_sample = sample.shape[axis]
  if num_sample < 2:
    raise ValueError("At least two observations are required.")
  return sample.std(axis=axis, ddof=1) / np.sqrt(num_sample)


def covariance_of_mean(sample: np.ndarray, sample_axis: int = 0) -> np.ndarray:
  """Estimate covariance matrices for one or more sample means.

  ``sample`` may have shape ``(n, d)`` for one mean or ``(n, m, d0)`` for
  ``m`` block means. The observation axis is moved to the front and the last
  axis is treated as the component axis. The result is ``W / [n(n - 1)]``.

  Parameters
  ----------
  sample : array_like
      Finite real observations with a final component axis. Other axes index
      separate mean vectors. At least two observations are required.
  sample_axis : int, default 0
      Observation axis, distinct from the component axis.

  Returns
  -------
  ndarray
      Shape ``(d, d)`` or ``(..., d, d)`` in squared input units. This is
      covariance of the estimated mean, not of individual observations.
      Positive definiteness is checked by the inference constructors, not here.

  Examples
  --------
  >>> import numpy as np
  >>> from heavily_right import covariance_of_mean, standard_error_of_mean
  >>> sample = np.array([[1., 2.], [2., 1.], [4., 5.]])
  >>> covariance = covariance_of_mean(sample)
  >>> assert np.allclose(covariance, np.cov(sample, rowvar=False, ddof=1) / 3)
  >>> assert np.allclose(np.diag(covariance), standard_error_of_mean(sample)**2)
  """
  sample = np.asarray(sample)
  if sample.ndim < 2:
    raise ValueError("Sample input must include observations and components.")

  sample = np.moveaxis(sample, sample_axis, 0)
  num_sample = sample.shape[0]
  if num_sample < 2:
    raise ValueError("At least two observations are required.")

  centered = sample - sample.mean(axis=0)
  scatter = np.einsum("n...i,n...j->...ij", centered, centered)
  return scatter / (num_sample * (num_sample - 1))

File: heavilyrightR/R/calibration.R
#' Standard error of a sample mean
#'
#' Uses the sample standard deviation with Bessel's correction (`n - 1`) and
#' divides by `sqrt(n)`.
#'
#' @param sample A numeric vector, matrix, or array.
#' @param margin Observation margin (ignored for a vector).
#' @return Componentwise standard errors.
#' @seealso [covariance_of_mean()]
#' @examples
#' x <- matrix(c(1, 2, 4, 7, 2, 5, 8, 11), ncol = 2)
#' standard_error_of_mean(x)
#' sqrt(diag(covariance_of_mean(x)))
#' standard_error_of_mean(x[, 1])
#' @export
standard_error_of_mean <- function(sample, margin = 1L) {
  .hr_assert_numeric(sample, "sample")
  d <- dim(sample)
  if (is.null(d)) {
    n <- length(sample)
    if (n < 2L) stop("at least two observations are required", call. = FALSE)
    return(stats::sd(sample) / sqrt(n))
  }
  margin <- as.integer(margin)
  if (length(margin) != 1L || margin < 1L || margin > length(d)) stop("invalid observation margin", call. = FALSE)
  n <- d[margin]
  if (n < 2L) stop("at least two observations are required", call. = FALSE)
  apply(sample, setdiff(seq_along(d), margin), stats::sd) / sqrt(n)
}

#' Covariance of one or more sample means
#'
#' Computes centered scatter divided by `n * (n - 1)`. This is covariance of
#' the sample mean, not covariance of individual observations. A matrix is
#' interpreted as observations by components. An array of shape `n x m x q`
#' returns `m` covariance matrices of shape `q x q`.
#'
#' @param sample Numeric matrix or three-dimensional array.
#' @param sample_margin Observation margin; currently must be the first margin.
#' @return A covariance matrix or a `q x q x m` array.
#' @seealso [standard_error_of_mean()], [study()]
#' @examples
#' x <- matrix(c(1, 2, 4, 7, 2, 5, 8, 11), ncol = 2)
#' covariance_of_mean(x)
#' stopifnot(isTRUE(all.equal(covariance_of_mean(x), stats::cov(x) / nrow(x))))
#' @export
covariance_of_mean <- function(sample, sample_margin = 1L) {
  .hr_assert_numeric(sample, "sample")
  d <- dim(sample)
  if (is.null(d) || length(d) < 2L || length(d) > 3L) {
    stop("sample must be a matrix or three-dimensional array", call. = FALSE)
  }
  if (sample_margin != 1L) stop("sample_margin must currently be 1", call. = FALSE)
  n <- d[1L]
  if (n < 2L) stop("at least two observations are required", call. = FALSE)
  if (length(d) == 2L) {
    centered <- sweep(sample, 2L, colMeans(sample), "-")
    return(crossprod(centered) / (n * (n - 1)))
  }
  m <- d[2L]
  q <- d[3L]
  out <- array(NA_real_, dim = c(q, q, m))
  for (j in seq_len(m)) {
    block <- sample[, j, , drop = FALSE]
    dim(block) <- c(n, q)
    centered <- sweep(block, 2L, colMeans(block), "-")
    out[, , j] <- crossprod(centered) / (n * (n - 1))
  }
  out
}
